docs/geps/gep-05.rst

```diff
@@ -0,0 +1,124 @@
+| Created    | 2022-01-08                                                       |
++------------+------------------------------------------------------------------+
+| Resolution | <url> (required for Accepted or Rejected or Withdrawn)           |
++------------+------------------------------------------------------------------+
+
+
+Abstract
+--------
+
+This GEP describes the implementation of optional rounding of variables in GETTSIM.
+
+
+Motivation and Scope
+--------------------
+
+For several taxes and transfers, German law specifies that these be rounded in specific
+ways. This leads to different use cases.
+
+1. Some applications require the exact, rounded, amount as specified in the law. This is
+   also helpful for creating test cases.
+2. Other applications require smooth functions and the non-rounding error does not
+   matter much.
+
+This document describes how we go about supporting both use cases.
+
+
+Implementation
+--------------
+
+GETTSIM allows for optional rounding of functions' results. Rounding parameters are
+specified in the ``.yaml``-files.
+
+The example below contains GETTSIM's encoding of the fact that the [XXX] law specifies
+that ``grundr_zuschlag_m`` be rounded to the nearest second decimal point.
+
+- The key has to be a function name, in this case ``grundr_zuschlag_m``
+- The parameter ``base`` determines the base to which the variables is rounded. It has
+  to be a floating point number.
+- The parameter ``direction`` has to be one of ``up``, ``down``, or ``nearest``.
+- The reference must contain the reference to the law, which specifies the rounding.
+
+The rounding parameters become part of the dictionary ``policy_params`` once
+``set_up_policy_environment`` is called in the same way as other policy parameters.
+
+.. code-block:: yaml
+
+       rounding:
+         grundr_zuschlag_m:
+           2020-01-01:
+             base: 0.01
+             direction: nearest
+             reference: Whatever law
+
+At the definition of the function ``grundr_zuschlag_m`` the decorator
+``add_rounding_spec_by_params_file`` indicates that the respective output should be
+rounded and in which parameter file the rounding parameters are specified.
+
+.. code-block:: python
+
+       @add_rounding_spec(params_file="ges_renten_vers")
+       def grundr_zuschlag_m(
+           grundr_zuschlag_vor_eink_anr_m: FloatSeries, anrechenbares_eink_gr_m: FloatSeries
+       ) -> FloatSeries:
+           ...
+           return out
+
+The decorator adds the attribute ``__rounding_parameter_file__`` to the function. When
+calling ``compute_taxes_and_transfers`` with ``rounding=True``, GETTSIM loops over all
+functions. If a function has a rounding attribute, the rounding is applied based on the
+specification in the argument ``policy_params``. In the example above, the rounding
+would be based on ``policy_params["ges_renten_vers"]["rounding"]["grundr_zuschlag_m"]``.
```

Comment:
Describe what happens if we specify only one but not the other.

- Specify in yaml but forget the decorator
- Specify decorator but forget in yaml
- If one thing has no consequences in terms of errors (probably the case for the first), specify a strategy for writing tests so as to minimize the chances of wrong assumptions regarding the behavior.